Tasks/DeployAzureResourceGroup/MGExtensionManager.ts

```diff
@@ -0,0 +1,132 @@
+    }
+
+    private setTaskResult() {
+        if (this.failureCount + this.successCount == this.vmCount) {
+            if (this.failureCount > 0) {
+                console.log("Machine group agent " + this.operation + " did not succeed on all VMs");
+            }
+            else {
+                console.log("Machine group agent " + this.operation + " succeeded on all VMs");
+            }
+            this.deferred.resolve("");
+        }
+    }
+
+    private postOperationCallBack = (error, result, request, response) => {
+        if (error) {
+            this.failureCount++;
+            this.errors += error.message;
+            this.errors += "\n";
+        } else {
+            this.successCount++;
+        }
+        this.setTaskResult();
+    }
+
+    public async installMGExtension() {
+        this.deferred = Q.defer<string>();
+        this.operation = "installation";
+        var listOfVms = await this.azureUtils.getVMDetails();
+        this.vmCount = listOfVms.length;
+        if(this.vmCount == 0){
+            this.deferred.resolve("");
+        }
+        for (var i = 0; i < listOfVms.length; i++) {
+            var vmName = listOfVms[i]["name"];
+            var extensionParameters = this.FormExtensionParameters(listOfVms[i], "enable");
+            console.log("Adding " + extensionParameters["extensionName"] + " extension to virtual machine " + vmName);
+            this.computeClient.virtualMachineExtensions.createOrUpdate(this.taskParameters.resourceGroupName, extensionParameters["vmName"], extensionParameters["extensionName"], extensionParameters["parameters"], this.postOperationCallBack);
+        }
+        return this.deferred.promise;
+    }
+
+    public async removeMGExtension() {
+        this.operation = "removal";
+        this.deferred = Q.defer<string>();
+        var listOfVms = await this.azureUtils.getVMDetails();
+        this.vmCount = listOfVms.length;
+        if(this.vmCount == 0){
+            this.deferred.resolve("");
+        }
+        for (var i = 0; i < listOfVms.length; i++) {
+            var vmName = listOfVms[i]["name"];
+            var extensionParameters = this.FormExtensionParameters(listOfVms[i], "uninstall");
+            console.log("Uninstalling " + extensionParameters["extensionName"] + " extension from virtual machine " + vmName);
+            this.computeClient.virtualMachineExtensions.deleteMethod(this.taskParameters.resourceGroupName, extensionParameters["vmName"], extensionParameters["extensionName"], this.postOperationCallBack);
+        }
+        return this.deferred.promise;
+    }
+
+    private FormExtensionParameters(virtualMachine, extensionAction) {
+        var vmId = virtualMachine["id"];
+        var vmName = virtualMachine["name"];
+        var vmOsType = virtualMachine["storageProfile"]["osDisk"]["osType"];
+        var vmLocation = virtualMachine["location"];
+
+        if (vmOsType == "Windows") {
+            var extensionName = "TeamServicesAgent";
```

Comment:
Variables have function scope in js. Declarations are hoisted to the top of the function